### PR TITLE
add moe context, moe utilities and refactor gradient handler

### PR DESCRIPTION
**Pull request by 1SAA**

### CODE REVIEW

File: colossalai/utils/moe.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
+import torch.nn as nn
+import torch.distributed as dist
+from colossalai.core import global_context as gpc, moe_context as moe_env
+from colossalai.context import ParallelMode
+from .common import is_using_ddp
+
+
+def get_moe_param_dict(model: nn.Module):
```
Comment (feifeibear) [inline colossalai/utils/moe.py:8]: -> dict[int, List[torch.nn.Parameters]]
develop a good habits of writing return value type...

File: tests/test_moe/short_test.py [hunk 2]
```diff
@@ -80,6 +80,7 @@ def run_routing(rank, world_size, port, rs=2, hidden_size=128, data_type=torch.f
     # print(f"linear weight gradient passed")
```
Comment (feifeibear) [inline tests/test_moe/short_test.py:80]: rm useless code.

File: colossalai/utils/moe.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+def get_moe_param_dict(model: nn.Module):
+    """Returns a parameter dictionary, the key of which is the expert parallel
+    size of every parameter. Since the parameters in data parallelism is replicated
+    in each GPU, we set their ep_size to 1.
+
+    :param model: A pyTorch nn.model from which we get dict
+    :type model: torch.nn.Module
+    """
+    param_dict = dict()
```
Comment (feifeibear) [inline colossalai/utils/moe.py:16]: rename it to epsize_param_dict will be good for you readers.

File: colossalai/utils/moe.py (hunk 1 above)
Comment (feifeibear) [reply colossalai/utils/moe.py:8]: rename it as get_moe_epsize_param_dict

Comment (1SAA) [reply colossalai/utils/moe.py]: done

File: tests/test_moe/short_test.py (hunk 2 above)
Comment (1SAA) [reply tests/test_moe/short_test.py]: done

File: colossalai/utils/moe.py (hunk 3 above)
Comment (1SAA) [reply colossalai/utils/moe.py]: done

**Review by feifeibear — APPROVED** [review]

File: colossalai/context/parallel_context.py [hunk 4]
```diff
@@ -407,12 +406,12 @@ def init_parallel_groups(self):
             # add this config to initialize later
             pg_init.append(dict(type=INITIALIZER_MAPPING[tensor_parallel_mode.lower()], **tensor_parallel_cfg))
 
-        # initialization for moe environment
-        if parallel_config is not None and 'moe' in parallel_config:
-            param = parallel_config['moe']
-            assert 'size' in param, "Moe model parallel size should be given"
-            moe_env.setup(param['size'])
-            pg_init.append(dict(type=INITIALIZER_MAPPING['moe']))
+        # deprecated for now (initialization for moe environment)
```
[the remark this answers is not in the document]
Comment (1SAA) [reply colossalai/context/parallel_context.py]: done
